### PR TITLE
Enable parallel collection in Dfs phase

### PR DESCRIPTION
**Pull request by cbuescher**
We now have the infrastructure to use parallel collection in the Dfs phase in place.
In order to actually use it, the IndexSearcher (ContextIndexSearcher in ES) needs 
to be initialized with a ThreadPoolExecutor that will enable parallel execution.
This change modifies the API so that the decision whether a phase should run
with parallelization enabled in the SearchService, when constructing the context
for each phases execution. Parallel execution is only enabled for the Dfs phase for
now, but can be enabled individually for each phase, also potentially taking into
account properties of the search request.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:509]: Given that out of a single search context instance, only either the parallel or the sequential search is used, it seems like a waste to create both all the time? 

I see a lot of different SearchService#createContext methods, maybe that is where we enable concurrency phase by phase?

**Review by javanna — COMMENTED** [review]
left some comments

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java:79]: No, left over from earlier version.

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:509]: I was under the impression we share the search context through the phases and want several phases to allow / disallow parallel execution individually. Also, afaik creating the searcher and the resources it consumes should be relatively light-weight. Please correct me if I'm wrong on that.
Will look into the different SearchService#createContext methods as options.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:509]: > I was under the impression we share the search context through the phases and want several phases to allow / disallow parallel execution individually.

Double check that, but we create an index searcher per shard level search at the lucene level. The index searcher is quite short-lived. In case we perform multiple queries or multiple knn searches, we do reuse the same instance, but that's the only situation. That is pretty much the same for the search context, I believe. 

> Also, afaik creating the searcher and the resources it consumes should be relatively light-weight. 

Sure, but it looks conceptually wrong to create an instance that we will never use.


Additional thought: we will have a need to tie the decision to a specific request (possibly based on info that are stored within ShardSearchRequest / features used in the SearchSourceBuilder?).

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java:212]: > this is a temporary fork of the lucene method

I keep forgetting that. the "supportsConcurrency" property isn't strictly needed, I introduced it here as a summary of all the different conditions this needs to be "true"

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java:120]: Looking at e.g. AbstractKnnVectorQuery#rewrite(Searcher), this should parallelize when AbstractKnnVectorQuery#sliceExecutor is set. So while this flag here isn't related to that, setting the executor on the ContextIndexSearch should propagate that to the super class. I made those changes now, regardless of whether we want to use this way or not.

File: server/src/main/java/org/elasticsearch/search/SearchService.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:449]: here we have the request at hand and could make additional decisions about parallelization based on its content

**Comment by cbuescher** [comment]
@javanna I pushed some updates that should address most of your previous comments. Seems like CI is happy now as well.

**Comment by javanna** [comment]
the approach looks reasonable to me. Note that we can't merge it using the SEARCH thread pool, we need first to introduce a separate thread pool. Working on that.

@iverase would you mind double checking the approach, does this align with what you had in mind? The goal is to start enabling concurrency for vector searches (dfs phase) separately from the query phase. Eventually, even within the query phase, I suspect we will need to make the decision looking at the request.

**Comment by cbuescher** [comment]
> we can't merge it using the SEARCH thread pool

Yes, that was my understanding.

**Comment by iverase** [comment]
I think this new logic is hardly been exercised on the tests because the minimum number of documents for concurrency is 50K and most of our test index far less documents. I think we should fix that (being able to force concurrency in tests regardless the number of documents) before merging this. We are adding a new behaviour that does not show on the tests, for example we should see deadlocks when using the SEARCH threadpool and CI is happy! 

I just realized the PR is stale and it does not contain the slicing logic we added this week so it is using the default lucene one which uses 250k as minimum documents for concurrency. 

As an exercise, you can override the `LeafSlice[] slices(List<LeafReaderContext> leaves)` method and create a slice per segment. We should see CI errors then.

In addition, I don't like to control the concurrency using the boolean parameter. This method is called from many places so it is very hard to understand when we are doing concurrency and when we are not. I was hoping to be able to decide it just by looking into the query. In my draft PR I was doing it by looking into `SearchSourceBuilder`:

https://github.com/elastic/elasticsearch/blob/5aa62af1d5aa6e6dcd2015d112a5b4e5170ba33d/server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java#L195

For example we should not be running concurrent if the aggregations on the query do not support it :
```
builder.aggregations() == null || builder.aggregations().supportsConcurrentExecution()
```

**Comment by iverase** [comment]
I went to the PR again and I understand you just want to enable it for one phase (not sure why this approach, is there any particular reason?). I would prefer to pass the phase instead of a boolean so we take the decision of paralelize in a single point.  

We agree that running with the profiler should not be done concurrent for now, should that be taken into consideration?

**Comment by cbuescher** [comment]
> just want to enable it for one phase (not sure why this approach, is there any particular reason?)

My understanding is we want to expedite this in Dfs phase first because we also would like to parallelize Knn search in particular. So this PR focuses just on that.

>  I would prefer to pass the phase instead of a boolean so we take the decision of paralelize in a single point.
> We agree that running with the profiler should not be done concurrent for now, should that be taken into consideration?

Good points, I will try to do that.

> I was hoping to be able to decide it just by looking into the query.

That certainly would be desirable If we decide that individual phases are not a factor in this decision I'm happy to push the checks down to e.g. the DefaultSearchContext

**Comment by cbuescher** [comment]
> We agree that running with the profiler should not be done concurrent for now, should that be taken into consideration?

Sorry, adding to that. I believe we have the infrastructure in place to profile things also in a concurrent setting, or at least we will have soon. In that case profiling or now shouldn't be a factor to enable/disable parallel execution anymore. But I will push a change that will make that decision easier to make in one spot.

**Comment by javanna** [comment]
I confirm that profiling should not be a factor in enabling/disabling concurrency, at least not for the DFS phase.

I confirm that we want to be able to enable concurrency for the DFS phase potentially before enabling it for the query phase. In the long run, looking at the request only should be sufficient, enabling it per phase is temporary to be able to enable concurrency phase by phase.

**Comment by javanna** [comment]
One more thought: as I make progress on the threading strategy, it becomes evident that we want to always offload work, even in the sequential case, to the executor. That has a direct consequence on this PR: we always need to have an executor set, and we need to have a way to disable parallelizing that is not based on whether the executor is set or not.

I expect to make progress on the thread pool front in the coming days that will make it easier to move forward with this PR too.

File: server/src/main/java/org/elasticsearch/search/SearchService.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:460]: unused

File: server/src/main/java/org/elasticsearch/search/query/QueryPhase.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:191]: This PR doesn't trigger this problem since enabling concurrency is limited to Dfs phase, but marking this as something to discuss for later.

**Comment by iverase** [comment]
The aggregation check looks good to me

**Review by javanna — COMMENTED** [review]
I reviewed this yesterday and forgot to submit the review :)

File: server/src/main/java/org/elasticsearch/search/SearchService.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:460]: Sure, changing back and adding comment to clarify this in the future

File: server/src/main/java/org/elasticsearch/search/query/QueryPhase.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:191]: I'll remove the TODO then

File: server/src/main/java/org/elasticsearch/search/SearchModule.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java:285]: Happy to change and will do so for now, though I though about something generic like this for a while and discarded it because "concurrency" in search is pretty general. e.g. if we also go into intra-segment concurrency later, will that be the same setting or something else? 
I will make the change though for now.

File: server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java:204]: I added this as a comment which should clarify this:
```
// The test collector manager throws an exception when the first segment gets collected.
// All documents in that slice count towards the "missing" docs in the later assertion.
```

File: server/src/main/java/org/elasticsearch/search/SearchService.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1094]: I switched the hard-to read oneliner to something more verbose and added tests.

**Comment by cbuescher** [comment]
@javanna thanks for the last review, I added changes that should address your comments.

File: server/src/main/java/org/elasticsearch/search/SearchModule.java
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java]: I think it's fine, and it's an escape hatch.

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:161]: just to be completely sure, can you replace AUTO_COMPLETE with something that does not compile? :)

File: server/src/main/java/org/elasticsearch/search/SearchModule.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/SearchModule.java:284]: can you update the constant name too?

File: server/src/main/java/org/elasticsearch/search/SearchService.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1097]: shall we look at aggs only when results type is query? Being paranoid :)

File: server/src/main/java/org/elasticsearch/search/SearchService.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1100]: I am actually wondering if it would not be better to default to false, and clearly enable it only for DFS above and later update for additional places like query that will want to enable concurrency. That may require commenting out some of this temporarily.

Also, are you not still looking at aggs even for the dfs phase? We don't want that, right?

File: server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java:205]: Thanks for clarifying the comments. I am a bit confused on what part of your change requires adapting this test. Sorry, I must be missing something.

**Review by javanna — COMMENTED** [review]
Left a few more cosmetic questions

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:161]: But leave it commented out I guess? Or make it so the PR doesn't compile?

File: server/src/main/java/org/elasticsearch/search/SearchService.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1097]: So far we would return in the condition above. I think this method needs adaption anyway when that part changed and we allow the other result types.

File: server/src/main/java/org/elasticsearch/search/SearchService.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1100]: I was under the impression we want that, but judging from your comment I'm probably wrong. Trying to answer here myself: since aggs are not collected in the dfs phase we probably don't need any of this for now. Again, to me this is a problem of having a moving target here, since we want to change this whole logic at some point but not right now. So I'm commenting out the aggs stuff completely for now, only leaving it here as a future plan.

File: server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java:205]: The assertion in L271 checks that visitDocs.get() + missingDocs.get(), equalTo(numDocs). So far the slicing overwrite we did here forces one segment for slice. With our own slicing logic now (`computeSlices(List<LeafReaderContext> leaves, int maxSliceNum, int minDocsPerSlice)`) we don't have that guarantee, we often end up with a few smaller segments in the first slice. Thats the one triggering the exception in L217. For the numbers to add up we not only need to count the docs from that one segment as "missing" but all docs in that slice.

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java]: leave it commented out. The plan is to merge your PR, and that's not possible if it does not compile :)

File: server/src/main/java/org/elasticsearch/search/SearchService.java
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: sounds good. The condition for aggs will only be relevant for the query phase. For DFS, you don't care if there are aggs, as they don't influence the execution of the DFS phase.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java:251]: Could you add a unit test to ContextIndexSearcherTests that verifies the discrepancy between the two slices method, and that either of the two is used when expected? That would mean e.g. running a knn vectory query, with the executor set to the searcher, and checking that rewrite creates one task per segment, while collection does not.

File: server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java
Comment (javanna) [reply server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java:205]: ok, and what is the value of changing how slices was overridden in this test?

**Review by javanna — APPROVED** [review]
left a suggestion for an additional test, LGTM otherwise

**Comment by cbuescher** [comment]
@javanna I added something to `ContextIndexSearcherTests` that checks that knn rewrite uses segment parallelism as long as there are enough threads available and extended the existing thread test with a check that we use one collector per slice. This should make sure the forked search path uses the "computeSlices" method.

File: server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java:205]: if you have an unbounded queue, like I think you have, the number of threads does not seem to matter for the assertions you have. tasks will queue up and be executed by less threads, but the number of execute calls will always be the same as the number of tasks. Rejections won't happen because the queue is unbounded.

File: server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java
Comment (javanna) [reply server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java:287]: I see, thanks, that makes sense.

File: server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java:236]: you may as well want to do a search here, and test how many tasks get created for that. Provide the rewritten query so that rewrite as part of search is a no-op.

File: server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java:205]: This is contrary to what I saw earlier when I was writing this test. It seems to be controlled by the [executors queue size](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java#L578). While the capacity of the `LinkedBlockingQueue` is unbounded, the actual queue size is not. I can set the maximumPoolSize of the executor to a higher value though, since this seems to be taken into account here as well.

**Comment by cbuescher** [comment]
@javanna I pushed some more changes to the tests

**Comment by javanna** [comment]
> I pushed some more changes to the tests

Still LGTM

**Pull request merged**
